apps/prairielearn/src/ee/lib/lti13.ts

```diff
@@ -211,3 +258,227 @@ export async function validateLti13CourseInstance(
+    return tokenSet.access_token;
+  } else {
+    const issuer = new Issuer(lti13_instance.issuer_params);
+    const client = new issuer.Client(lti13_instance.client_params, lti13_instance.keystore);
+
+    tokenSet = await client.grant({
+      grant_type: 'client_credentials',
+      scope: TOKEN_SCOPES.join(' '),
+    });
+
+    // Store the token for reuse
+
+    const expires_at = tokenSet.expires_at ? tokenSet.expires_at * 1000 : Date.now();
+
+    await queryAsync(sql.update_token, {
+      lti13_instance_id,
+      tokenSet,
+      expires_at: new Date(expires_at),
+    });
+    return tokenSet.access_token;
+  }
+}
+
+export async function get_lineitems(instance: Lti13CombinedInstance) {
+  if (instance.lti13_course_instance.lineitems == null) {
+    throw new Error('Lineitems not defined');
+  }
+  const token = await access_token(instance.lti13_instance.id);
+  const lineitems: Lti13Lineitem[] = await fetchRetry(instance.lti13_course_instance.lineitems, {
+    method: 'GET',
+    headers: { Authorization: `Bearer ${token}` },
+  });
+
+  await fetchRetry('https://httpbin.org/status/403');
+
+  return lineitems;
+}
+
+/*
+ * Canvas supports an API call to get a single lineitem.
+ * https://canvas.instructure.com/doc/api/line_items.html#method.lti/ims/line_items.show
+ * But in testing Dave discovered its output might not be the same as
+ * lineitems (plural).
+ *
+ * Instead of a get_lineitem() function, let's get_lineitems and filter.
+ */
+
+/////////////////////////////////////////////////////////////////////////////////////////////////
+export async function sync_lineitems(instance: Lti13CombinedInstance, job: ServerJob) {
+  job.info(
+    `Polling for external assignments from ${instance.lti13_instance.name} ${instance.lti13_course_instance.context_label}`,
+  );
+  const lineitems = await get_lineitems(instance);
+  job.info(`Found ${lineitems.length}`);
+
+  await runInTransactionAsync(async () => {
+    await queryAsync(sql.create_lineitems_temp, {});
+
+    for (const item of lineitems) {
+      job.info(`* ${item.label}`);
+
+      await queryAsync(sql.insert_lineitems_temp, {
+        lti13_course_instance_id: instance.lti13_course_instance.id,
+        lineitem_id: item.id,
+        lineitem: JSON.stringify(item),
+      });
+    }
+
+    const output = await queryRow(
+      sql.sync_lti13_lineitems,
+      {
+        lti13_course_instance_id: instance.lti13_course_instance.id,
+      },
+      z.object({
+        updated: z.string(),
+        deleted: z.string(),
+      }),
+    );
+
+    job.info(
+      `\nSummary of PrairieLearn changes: ${output.updated} updated, ${output.deleted} deleted.`,
+    );
+    job.info('Done.');
+  });
+}
+
+export async function create_and_link_lineitem(
+  instance: any,
+  job: ServerJob,
+  assessment: {
+    label: string;
+    id: string;
+    url: string;
+  },
+) {
+  const createBody: Lti13Lineitem = {
+    id: 'new_lineitem', // will be ignored/overwritten by the LMS platform
```

Comment:
It was to make the typing work. TS didn't like me making a `Lti13Lineitem` object without an ID to send to the POST API.
https://canvas.instructure.com/doc/api/line_items.html#method.lti/ims/line_items.create

Reply:
I'd recommend changing the types to match reality, rather than code to match the types. You can type this as `Omit<Lti13LineItem, 'id'>` or something like that.

Reply:
Addressed in 46c87e028